File: analyses/divorce-chaebol-awakening/_working/validate-final.mjs
import { createHash } from "node:crypto";
import { readFileSync, statSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const workDir = dirname(fileURLToPath(import.meta.url));
const outputDir = dirname(workDir);
const receipt = JSON.parse(readFileSync(join(outputDir, "source_receipt.json"), "utf8"));

const expectedChapterHeader = "sequence,visible_label,title,start_line,end_line,entry_state,reader_promise,protagonist_goal,action,resistance_or_cost,turn_or_reveal,paid_reward,state_change_axis,state_change,ending_hook,closed_loops,opened_loops,arc_id,confidence";
const expectedArcHeader = "arc_id,arc_name,start_sequence,end_sequence,start_label,end_label,episode_count,main_characters,main_locations,concrete_premise,central_question,promise,pressure_escalation,mid_turn,concrete_payoff,relationship_change,status_or_ability_change,residual_cost,next_arc_bridge,boundary_signals,confidence";
const expectedPacingHeader = "sequence,visible_label,arc_id,arc_phase,concrete_event,tension_1_10,reward_1_10,hook_1_10,information_weight,action_weight,relationship_weight,emotion_weight,material_or_status_weight,pacing_note";

function assert(condition, message) {
  if (!condition) throw new Error(message);
}

function parseCsv(text) {
  const records = [];
  let record = [];
  let field = "";
  let quoted = false;
  for (let index = 0; index < text.length; index += 1) {
    const char = text[index];
    if (quoted) {
      if (char === '"' && text[index + 1] === '"') {
        field += '"';
        index += 1;
      } else if (char === '"') {
        quoted = false;
      } else {
        field += char;
      }
    } else if (char === '"') {
      quoted = true;
    } else if (char === ",") {
      record.push(field);
      field = "";
    } else if (char === "\n") {
      record.push(field.replace(/\r$/, ""));
      records.push(record);
      record = [];
      field = "";
    } else {
      field += char;
    }
  }
  assert(!quoted, "unterminated CSV quote");
  if (field || record.length) {
    record.push(field.replace(/\r$/, ""));
    records.push(record);
  }
  const header = records.shift();
  return { header, rows: records.map((cells) => Object.fromEntries(header.map((key, index) => [key, cells[index] ?? ""]))) };
}

const source = readFileSync(receipt.sourcePath);
const sourceText = source.toString("utf8").replaceAll("\r\n", "\n");
const sourceLines = sourceText.split("\n");
if (sourceLines.at(-1) === "") sourceLines.pop();
const sourceHash = createHash("sha256").update(source).digest("hex");
assert(sourceHash === receipt.sourceSha256, `source hash mismatch: ${sourceHash}`);
assert(sourceHash === "09ca4bb08e88f7791e38bc33c393ee1732af32823d7ed39a1c7026f4b7e37908", "source hash differs from order");

const markers = sourceLines
  .map((line, index) => ({ text: line, line: index + 1 }))
  .filter((entry) => entry.text.startsWith("ⓚ"));
assert(markers.length === 220, `expected 220 markers, got ${markers.length}`);
assert(markers[0].text === receipt.firstMarker, "first marker mismatch");
assert(markers.at(-1).text === receipt.lastMarker, "last marker mismatch");

const chapterRaw = readFileSync(join(outputDir, "chapter_map.csv"), "utf8");
const arcRaw = readFileSync(join(outputDir, "arc_map.csv"), "utf8");
const pacingRaw = readFileSync(join(outputDir, "arc_pacing.csv"), "utf8");
assert(chapterRaw.split(/\r?\n/, 1)[0] === expectedChapterHeader, "chapter_map.csv header mismatch");
assert(arcRaw.split(/\r?\n/, 1)[0] === expectedArcHeader, "arc_map.csv header mismatch");
assert(pacingRaw.split(/\r?\n/, 1)[0] === expectedPacingHeader, "arc_pacing.csv header mismatch");

const chapterCsv = parseCsv(chapterRaw);
const arcCsv = parseCsv(arcRaw);
const pacingCsv = parseCsv(pacingRaw);
assert(chapterCsv.rows.length === 220, `chapter_map.csv has ${chapterCsv.rows.length} data rows`);
assert(pacingCsv.rows.length === 220, `arc_pacing.csv has ${pacingCsv.rows.length} data rows`);
assert(arcCsv.rows.length > 0, "arc_map.csv has no data rows");

const essentialChapterFields = expectedChapterHeader.split(",").filter((field) => !["closed_loops", "opened_loops"].includes(field));
for (const [index, row] of chapterCsv.rows.entries()) {
  const sequence = index + 1;
  assert(Number(row.sequence) === sequence, `chapter row ${sequence}: sequence mismatch`);
  const expectedLabel = sequence === 220 ? "220화(완결화)" : `${sequence}화`;
  assert(row.visible_label === expectedLabel, `chapter row ${sequence}: visible label mismatch`);
  assert(Number(row.start_line) === markers[index].line, `chapter row ${sequence}: start line mismatch`);
  const expectedEnd = index + 1 < markers.length ? markers[index + 1].line - 1 : sourceLines.length;
  assert(Number(row.end_line) === expectedEnd, `chapter row ${sequence}: end line mismatch`);
  assert(row.title === markers[index].text.slice(1), `chapter row ${sequence}: title differs from source marker`);
  for (const field of essentialChapterFields) assert(row[field].trim(), `chapter row ${sequence}: empty ${field}`);
}

const uniqueFields = ["protagonist_goal", "resistance_or_cost", "paid_reward", "state_change", "ending_hook"];
for (const field of uniqueFields) {
  const unique = new Set(chapterCsv.rows.map((row) => row[field])).size;
  assert(unique >= 210, `${field} has only ${unique} unique values; possible blanket duplication`);
}

let expectedStart = 1;
const arcById = new Map();
for (const [index, arc] of arcCsv.rows.entries()) {
  const expectedId = `ARC-${String(index + 1).padStart(3, "0")}`;
  assert(arc.arc_id === expectedId, `Arc ${index + 1}: id mismatch`);
  const start = Number(arc.start_sequence);
  const end = Number(arc.end_sequence);
  assert(start === expectedStart, `${arc.arc_id}: range is not contiguous`);
  assert(end >= start && end <= 220, `${arc.arc_id}: invalid end`);
  assert(Number(arc.episode_count) === end - start + 1, `${arc.arc_id}: episode count mismatch`);
  const expectedEndLabel = end === 220 ? "220화(완결화)" : `${end}화`;
  assert(arc.start_label === `${start}화` && arc.end_label === expectedEndLabel, `${arc.arc_id}: label range mismatch`);
  assert(arc.boundary_signals.split(" | ").filter(Boolean).length >= 2, `${arc.arc_id}: fewer than two boundary signals`);
  for (const [field, value] of Object.entries(arc)) assert(value.trim(), `${arc.arc_id}: empty ${field}`);
  arcById.set(arc.arc_id, { start, end });
  expectedStart = end + 1;
}
assert(expectedStart === 221, `Arc coverage ends at ${expectedStart - 1}`);

for (const [index, row] of chapterCsv.rows.entries()) {
  const sequence = index + 1;
  const arc = arcById.get(row.arc_id);
  assert(arc && sequence >= arc.start && sequence <= arc.end, `chapter ${sequence}: wrong or missing Arc ${row.arc_id}`);
  const pacing = pacingCsv.rows[index];
  assert(Number(pacing.sequence) === sequence, `pacing row ${sequence}: sequence mismatch`);
  assert(pacing.visible_label === row.visible_label, `pacing row ${sequence}: label mismatch`);
  assert(pacing.arc_id === row.arc_id, `pacing row ${sequence}: Arc mismatch`);
  for (const field of ["tension_1_10", "reward_1_10", "hook_1_10"]) {
    const value = Number(pacing[field]);
    assert(Number.isInteger(value) && value >= 1 && value <= 10, `pacing row ${sequence}: ${field} out of range`);
  }
  for (const field of ["information_weight", "action_weight", "relationship_weight", "emotion_weight", "material_or_status_weight"]) {
    assert(["low", "medium", "high"].includes(pacing[field]), `pacing row ${sequence}: invalid ${field}`);
  }
  for (const [field, value] of Object.entries(pacing)) assert(value.trim(), `pacing row ${sequence}: empty ${field}`);
}

const requiredDocs = {
  "project_bible.md": 20_000,
  "arc_atlas.md": 60_000,
  "inkos_usage_and_gap_report.md": 20_000,
  "free_improvements_report.md": 10_000,
  "completion_receipt.md": 2_000,
};
for (const [filename, minimumBytes] of Object.entries(requiredDocs)) {
  const size = statSync(join(outputDir, filename)).size;
  assert(size >= minimumBytes, `${filename} is too small: ${size} bytes`);
}

const result = {
  sourceSha256: sourceHash,
  sourceLines: sourceLines.length,
  markerCount: markers.length,
  chapterRows: chapterCsv.rows.length,
  arcRows: arcCsv.rows.length,
  pacingRows: pacingCsv.rows.length,
  chapterRange: `${chapterCsv.rows[0].visible_label}~${chapterCsv.rows.at(-1).visible_label}`,
  lineRange: `${chapterCsv.rows[0].start_line}~${chapterCsv.rows.at(-1).end_line}`,
  uniqueChapterFields: Object.fromEntries(uniqueFields.map((field) => [field, new Set(chapterCsv.rows.map((row) => row[field])).size])),
  status: "PASS",
};
console.log(JSON.stringify(result, null, 2));

File: analyses/doksik-chaebol3/.work/archive/2026-08-15-zero-loss/legacy_flat_work/validate_final.rb
#!/usr/bin/env ruby

require "csv"
require "digest"
require "json"

ROOT = File.expand_path("..", __dir__)
SOURCE = File.expand_path(
  "../../private_sources/korean_webnovel_corpus/서오/독식하는 재벌 3세_서오_합본.txt",
  ROOT
)
EXPECTED_SHA = "66f3e7df3123343c14a7134108dc59940511f40372493a549cba8ae709df4b45"

CHAPTER_HEADER = %w[
  sequence visible_label title start_line end_line entry_state reader_promise
  protagonist_goal action resistance_or_cost turn_or_reveal paid_reward
  state_change_axis state_change ending_hook closed_loops opened_loops arc_id
  confidence
].freeze
PACING_HEADER = %w[
  sequence visible_label arc_id arc_phase concrete_event tension_1_10
  reward_1_10 hook_1_10 information_weight action_weight relationship_weight
  emotion_weight material_or_status_weight pacing_note
].freeze
ARC_HEADER = %w[
  arc_id arc_name start_sequence end_sequence start_label end_label episode_count
  main_characters main_locations concrete_premise central_question promise
  pressure_escalation mid_turn concrete_payoff relationship_change
  status_or_ability_change residual_cost next_arc_bridge boundary_signals confidence
].freeze

failures = []
check = lambda do |condition, message|
  failures << message unless condition
end

receipt = JSON.parse(File.read(File.join(ROOT, "source_receipt.json")))
sha = Digest::SHA256.file(SOURCE).hexdigest
check.call(sha == EXPECTED_SHA, "source SHA mismatch")
check.call(receipt.fetch("sourceSha256") == sha, "receipt SHA mismatch")

lines = File.readlines(SOURCE, chomp: true)
markers = []
lines.each_with_index do |line, index|
  next unless line.start_with?("ⓚ")
  match = line.match(/\Aⓚ(\d+)(화)?(?:\.|\s)\s*(.*?)\s*\z/)
  unless match
    failures << "unparsed marker at line #{index + 1}: #{line}"
    next
  end
  visible = "#{match[1]}#{match[2]}"
  markers << {
    "sequence" => markers.length + 1,
    "visible_label" => visible,
    # Two source markers (356 and 357) end in U+00A0. Preserve the marker and
    # line boundary, but compare the human-visible title after terminal-space
    # normalization; chapter_map records no invisible title suffix.
    "title" => match[3].sub(/[\s\u00A0]+\z/, ""),
    "start_line" => index + 1
  }
end
markers.each_with_index do |marker, index|
  marker["end_line"] = index + 1 < markers.length ? markers[index + 1]["start_line"] - 1 : lines.length
end
check.call(markers.length == 751, "source marker count #{markers.length}, expected 751")
check.call(markers.map { |m| m.fetch("visible_label").to_i } == (1..751).to_a, "visible numbers not 1..751")

chapter = CSV.read(File.join(ROOT, "chapter_map.csv"), headers: true)
pacing = CSV.read(File.join(ROOT, "arc_pacing.csv"), headers: true)
arcs = CSV.read(File.join(ROOT, "arc_map.csv"), headers: true)
check.call(chapter.headers == CHAPTER_HEADER, "chapter header mismatch")
check.call(pacing.headers == PACING_HEADER, "pacing header mismatch")
check.call(arcs.headers == ARC_HEADER, "arc header mismatch")
check.call(chapter.length == 751, "chapter rows #{chapter.length}, expected 751")
check.call(pacing.length == 751, "pacing rows #{pacing.length}, expected 751")
check.call(arcs.length == 97, "Arc rows #{arcs.length}, expected 97")

chapter.each_with_index do |row, index|
  marker = markers[index]
  sequence = index + 1
  check.call(row["sequence"].to_i == sequence, "chapter sequence mismatch at #{sequence}")
  check.call(row["visible_label"] == marker["visible_label"], "visible label mismatch at #{sequence}: #{row['visible_label']} != #{marker['visible_label']}")
  check.call(row["title"] == marker["title"], "title mismatch at #{sequence}: #{row['title']} != #{marker['title']}")
  check.call(row["start_line"].to_i == marker["start_line"], "start line mismatch at #{sequence}")
  check.call(row["end_line"].to_i == marker["end_line"], "end line mismatch at #{sequence}")
  CHAPTER_HEADER.each { |field| check.call(!row[field].to_s.strip.empty?, "blank chapter #{field} at #{sequence}") }
  confidence = row["confidence"].to_f
  check.call(confidence.positive? && confidence <= 1.0, "chapter confidence #{row['confidence']} at #{sequence}")

  pacing_row = pacing[index]
  check.call(pacing_row["sequence"] == row["sequence"], "pacing sequence mismatch at #{sequence}")
  check.call(pacing_row["visible_label"] == row["visible_label"], "pacing label mismatch at #{sequence}")
  check.call(pacing_row["arc_id"] == row["arc_id"], "pacing arc mismatch at #{sequence}")
  PACING_HEADER.each { |field| check.call(!pacing_row[field].to_s.strip.empty?, "blank pacing #{field} at #{sequence}") }
  %w[tension_1_10 reward_1_10 hook_1_10].each do |field|
    value = pacing_row[field].to_i
    check.call((1..10).cover?(value), "pacing #{field}=#{pacing_row[field]} at #{sequence}")
  end
end

coverage = []
seen_arc_ids = {}
arcs.each_with_index do |row, index|
  start_sequence = row["start_sequence"].to_i
  end_sequence = row["end_sequence"].to_i
  check.call(start_sequence <= end_sequence, "reversed Arc #{row['arc_id']}")
  check.call(row["episode_count"].to_i == end_sequence - start_sequence + 1, "episode_count mismatch #{row['arc_id']}")
  check.call(!seen_arc_ids.key?(row["arc_id"]), "duplicate Arc row #{row['arc_id']}")
  seen_arc_ids[row["arc_id"]] = true
  ARC_HEADER.each { |field| check.call(!row[field].to_s.strip.empty?, "blank Arc #{field} at #{row['arc_id']}") }
  arc_confidence = row["confidence"].to_f
  check.call(arc_confidence.positive? && arc_confidence <= 1.0, "Arc confidence #{row['confidence']} at #{row['arc_id']}")
  check.call(row["arc_name"] !~ /\A(?:S\d|Arc|구간)/, "non-event Arc name at #{row['arc_id']}")
  %w[main_characters main_locations concrete_premise concrete_payoff next_arc_bridge].each do |field|
    check.call(row[field].to_s.length >= 8, "thin Arc surface #{field} at #{row['arc_id']}")
  end
  signal_count = row["boundary_signals"].split(/[;·\/]/).map(&:strip).reject(&:empty?).length
  check.call(signal_count >= 2, "fewer than two boundary signals at #{row['arc_id']}")
  coverage.concat((start_sequence..end_sequence).to_a)
  range_rows = chapter[(start_sequence - 1)..(end_sequence - 1)] || []
  check.call(range_rows.all? { |episode| episode["arc_id"] == row["arc_id"] }, "chapter Arc range mismatch #{row['arc_id']}")
  chapter_arc_count = chapter.count { |episode| episode["arc_id"] == row["arc_id"] }
  check.call(chapter_arc_count == range_rows.length, "non-contiguous chapter Arc id #{row['arc_id']}")
  if index.positive?
    previous = arcs[index - 1]
    check.call(previous["end_sequence"].to_i + 1 == start_sequence, "Arc gap/overlap before #{row['arc_id']}")
  end
end
check.call(coverage == (1..751).to_a, "Arc coverage not exact 1..751")

required = %w[
  source_receipt.json chapter_map.csv project_bible.md arc_map.csv arc_pacing.csv
  arc_atlas.md inkos_usage_and_gap_report.md free_improvements_report.md
  completion_receipt.md
]
required.each do |name|
  path = File.join(ROOT, name)
  check.call(File.file?(path) && File.size(path).positive?, "missing or empty #{name}")
end
analysis_files = Dir.glob(File.join(ROOT, "*.{md,csv,json}"))
analysis_files.each do |path|
  content = File.read(path)
  check.call(!content.include?("다이몬"), "surface typo 다이몬 in #{File.basename(path)}")
end

atlas = File.read(File.join(ROOT, "arc_atlas.md"))
arc_ids = arcs.map { |row| row["arc_id"] }
arc_ids.each { |arc_id| check.call(atlas.include?("## #{arc_id} ·"), "atlas missing #{arc_id}") }
atlas_episodes = atlas.scan(/^- (\d+)화 ·/).flatten.map(&:to_i)
check.call(atlas_episodes.length == 751, "atlas episode beats #{atlas_episodes.length}, expected 751")
check.call(atlas_episodes.uniq.sort == (1..751).to_a, "atlas episode beat coverage mismatch")

if failures.empty?
  puts "PASS"
  puts "sha256=#{sha}"
  puts "markers=#{markers.length}"
  puts "chapter_rows=#{chapter.length}"
  puts "arc_rows=#{arcs.length}"
  puts "pacing_rows=#{pacing.length}"
  puts "atlas_episode_beats=#{atlas_episodes.length}"
else
  warn failures.uniq.join("\n")
  exit 1
end
